src/MutableMap/MutableMap.dfy

```diff
@@ -0,0 +1,47 @@
+/*******************************************************************************
+*  Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT
+*******************************************************************************/
+
+// RUN: %dafny /compile:0 "%s"
+
+module {:extern "MutableMap"} {:options "/functionSyntax:4"} MutableMap {
+  class {:extern "MutableMap"} MutableMap<K(==),V(==)> {
+    function {:extern "content"} content(): map<K, V>
+      reads this
+
+    constructor {:extern "MutableMap"} ()
+      ensures this.content() == map[]
+
+    method {:extern "put"} Put(k: K, v: V)
+      modifies this
+      ensures this.content() == old(this.content())[k := v]    
+
+    function {:extern "keys"} Keys(): (keys: set<K>)
+      reads this
+      ensures keys == this.content().Keys
+
+    function {:extern "values"} Values(): (values: set<V>)
+      reads this
+      ensures values == this.content().Values
+
+    function {:extern "items"} Items(): (items: set<(K,V)>)
+      reads this
+      ensures items == this.content().Items
+      ensures items == set k | k in this.content().Keys :: (k, this.content()[k])
```

Comment:
This is technically redundant, did you find it necessary? It feels like it should be a separate lemma from Maps.dfy instead.

Reply:
At some point it seemed like I need both ensures clauses to prove the assert statements in `MutableMapExamples.dfy`. But now they go through with just the second. I will remove the first one.

Reply:
Even if an ensures clause is not needed to make the example proofs go through, I would tend to keep it for a library if it seemed it might save the prover some work or make a fact know that a user proof might need. The example program has lots of asserts so it might inadvertently guide the prover where a a less sparsely annotated program might not. I'd be good with keeping it. Your call.